torch_xla/csrc/xla_graph_executor.cpp

```diff
@@ -502,7 +505,10 @@ XLAGraphExecutor::SyncTensorCollection XLAGraphExecutor::CollectSyncTensors(
                                   tsl::profiler::TraceMeLevel::kInfo);
   runtime::util::Unique<torch::lazy::BackendDevice> unique_device;
   for (size_t i = 0; i < tensors.size(); ++i) {
-    unique_device.set(tensors[i]->GetDevice());
+    // TODO(JackCaoG): all tensors should be on spmd:0 in SPMD mode.
+    if (tensors[i]->GetDevice().toString() != "SPMD:0") {
```

Comment:
So currently we allow SPMD:0 and TPU:0 to coexist? I wonder what's the deal with the code later on that deals with unique_device?

Reply:
I will move all tensors on `SPMD:0` under spmd context. Planning on work on that later this week. Currently the code that go through
```
tensor.to(xla_device)
```
will be on `TPU:0` and tensors go through
```
# data loader
`xm.send_cpu_data_to_device`
```
will be on `SPMD:0`, this is really confusing and make logic nasty.

Reply:
Right, but we have code in this method later on use unique_device? And it looks like it's important: `coll.device = *unique_device;` How does it work today?

And then if SPMD:0 and TPU:0 really coexist in this method, shouldn't `unique_device.set` crash?

Reply:
oh, so here is where things become messy. What happened is that we will pass `TPU:0` as the device to `Compile` and `ScheduleSyncTensorgraph` on xla_graph_executor. `ScheduleSyncTensorsGraph` will ignore the device and check if graph is sharded and execute on all device if that's the case

https://github.com/pytorch/xla/blob/master/torch_xla/csrc/xla_graph_executor.cpp#L999-L1000

It is a very broken design now. Different function is checking different thing. I want to unify all those later.

Reply:
Err... whatever, I will approve that and then wait for your redesign.